### PR TITLE
🩹 fix: Timeout middleware not enforced

### LINKED ISSUE #3671: 🐛 [Bug]: timeout is not actually enforced
### Bug Description

Doesn't this just check for an error afterward? From what I see, the timeout doesn't actually stop the handler from running. It also doesn't pass the cancellation down the call stack, so inside the handler there's no way to check if the request has timed out.

https://github.com/gofiber/fiber/blob/c309d46c3a0fcf6f3ca49f7e2aa8e3e8b4ae8e48/middleware/timeout/timeout.go#L25-L28

### How to Reproduce

1. Go to '....'
3. Do '....'
4. See '....'

### Expected Behavior

Context(tCtx) with cancelation is passed down

### Fiber Version

v3.0.0-beta.5

### Code Snippet (optional)

```go

```

### Checklist:

- [x] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/main/.github/CODE_OF_CONDUCT.md).
- [x] I have checked for existing issues that describe my problem prior to opening this one.
- [x] I understand that improperly formatted bug reports may be closed without explanation.

### PR DESCRIPTION
**Pull request by edvardsanta**
# Description

This PR refactors the timeout middleware to properly enforce request timeouts by:

1. Wrapping the handler execution in a goroutine.
2 Using a `select` statement to handle three possible outcomes:

    - Handler completes successfully.
    - Handler panics (panic is propagated).
    - Timeout occurs, triggering cfg.OnTimeout if set, or returning fiber.ErrRequestTimeout.. 

This change guarantees that:
- The timeout is respected in real time.
- Handlers have the opportunity to observe context cancellation if needed.
- Panics are safely propagated without losing stack traces.

Fixes #3671

## Changes introduced

List the new features or adjustments introduced in this pull request. Provide details on benchmarks, documentation updates, changelog entries, and if applicable, the migration guide.

- [ ] Benchmarks: Describe any performance benchmarks and improvements related to the changes.
- [ ] Documentation Update: Detail the updates made to the documentation and links to the changed files.
- [ ] Changelog/What's New: Include a summary of the additions for the upcoming release notes.
- [ ] Migration Guide: If necessary, provide a guide or steps for users to migrate their existing code to accommodate these changes.
- [ ] API Alignment with Express: Explain how the changes align with the Express API.
- [ ] API Longevity: Discuss the steps taken to ensure that the new or updated APIs are consistent and not prone to breaking changes.
- [ ] Examples: Provide examples demonstrating the new features or changes in action.

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [x] Enhancement (improvement to existing features and functionality)
- [ ] Documentation update (changes to documentation)
- [ ] Performance improvement (non-breaking change which improves efficiency)
- [ ] Code consistency (non-breaking change which improves code reliability and robustness)

## Checklist

Before you submit your pull request, please make sure you meet these requirements:

- [ ] Followed the inspiration of the Express.js framework for new functionalities, making them similar in usage.
- [] Conducted a self-review of the code and provided comments for complex or critical parts.
- [ ] Updated the documentation in the `/docs/` directory for [Fiber's documentation](https://docs.gofiber.io/).
- [ ] Added or updated unit tests to validate the effectiveness of the changes or new features.
- [ ] Ensured that new and existing unit tests pass locally with the changes.
- [ ] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community.
- [ ] Aimed for optimal performance with minimal allocations in the new code.
- [ ] Provided benchmarks for the new code to analyze and improve upon.

## Commit formatting

Please use emojis in commit messages for an easy way to identify the purpose or intention of a commit. Check out the emoji cheatsheet here: [CONTRIBUTING.md](https://github.com/gofiber/fiber/blob/main/.github/CONTRIBUTING.md#pull-requests-or-commits)

### CODE REVIEW

**Comment by edvardsanta** [comment]
Hello, @ReneWerner87, @gaby ,@own2pwn, @sixcolors.

I’m exploring some edge cases around the **Timeout** middleware and would like your thoughts on expected behavior. Specifically, I’m wondering how Fiber should behave when a panic occurs in the handler. I have written two tests cases:
**Handler panics after timeout**:
```go
handler: func(_ fiber.Ctx) error {
      time.Sleep(50 * time.Millisecond)
      panic("panic after timeout")
}
```
**Custom OnTimeout handler panics**:
``` go
handler: func(_ fiber.Ctx) error {
		time.Sleep(50 * time.Millisecond)
		return nil
	},
	config: Config{
		Timeout: 10 * time.Millisecond,
		OnTimeout: func(_ fiber.Ctx) error {
			panic("custom panic on timeout")
		},
},
```

My questions:

1. In the case of a panic after the timeout has already triggered, should Fiber still return a 408 Request Timeout, or should the panic propagate?

2. For a panic inside a custom **OnTimeout** handler, do we have an expected status code, or should the panic be allowed to crash the app?

3. Any recommendations on handling these panics safely in production while maintaining the correct timeout semantics?

I’m not sure what the “correct” behavior should be here, so I’d like to align it first before to open the pr

**Comment by gaby** [comment]
@edvardsanta 

1. Return 408
2. Probably 408 too
3. Add tests for them, testify has a way of testing these

**Comment by gaby** [comment]
@edvardsanta Any updates on this?

**Comment by edvardsanta** [comment]
> @edvardsanta Any updates on this?

I’ll get back to this later today

File: middleware/timeout/timeout.go [hunk 1]
```diff
@@ -31,16 +31,26 @@ func New(h fiber.Handler, config ...Config) fiber.Handler {
...
-		if errors.Is(tCtx.Err(), context.DeadlineExceeded) && err == nil {
-			if cfg.OnTimeout != nil {
-				return cfg.OnTimeout(ctx)
-			}
+		select {
+		case err := <-done:
+			return err
+		case <-panicChan:
+			return fiber.ErrRequestTimeout
+		case <-tCtx.Done():
```
[the remark this answers is not in the document]
Comment (edvardsanta) [reply middleware/timeout/timeout.go]: Yeah, i noticed that in `TestTimeout_CustomHandler`, the handler returns `context.DeadlineExceeded `under `/custom-handler.`
Because of this condition:
```go
if err != nil && (errors.Is(err, context.DeadlineExceeded) || (len(cfg.Errors) > 0 && isCustomError(err, cfg.Errors)))
```
it ends up always expecting an error.
However, there’s also a case where the context deadline can be reached before the blocking condition inside the handler triggers an error

File: middleware/timeout/timeout.go [hunk 2]
```diff
@@ -31,16 +31,26 @@ func New(h fiber.Handler, config ...Config) fiber.Handler {
...
 
-		if errors.Is(tCtx.Err(), context.DeadlineExceeded) && err == nil {
-			if cfg.OnTimeout != nil {
-				return cfg.OnTimeout(ctx)
-			}
+		select {
+		case err := <-done:
+			return err
+		case <-panicChan:
+			return fiber.ErrRequestTimeout
```
[the remark this answers is not in the document]
Comment (edvardsanta) [reply middleware/timeout/timeout.go:80]: https://github.com/gofiber/fiber/pull/3680#issuecomment-3194428122

File: middleware/timeout/timeout.go [hunk 3]
```diff
@@ -21,29 +20,50 @@
...
+			}()
+			done <- runHandler(ctx, h, cfg)
+		}()
 
-		if errors.Is(tCtx.Err(), context.DeadlineExceeded) {
+		select {
+		case err := <-done:
+			return err
+		case _ = <-panicChan:
+			return fiber.ErrRequestTimeout
```
[the remark this answers is not in the document]
Comment (edvardsanta) [reply middleware/timeout/timeout.go]: https://github.com/gofiber/fiber/pull/3680#issuecomment-3194428122

**Comment by ReneWerner87** [comment]
@edvardsanta can you check https://github.com/gofiber/fiber/pull/4009
